site/bps/BP-38-bookie-endpoint-discovery.md

```diff
@@ -0,0 +1,149 @@
+
+### Public Interfaces
+
+On the Registration API, we introduce a new data structure that describes the services
+exposed by a Bookie:
+
+```
+inteface BookieServiceInfo {
+    class Endpoint {
+        String name;
+        String hostname;
+        int port;
+        String protocol; // "bookie-rpc", "http", "https"....
+        String[] auth; // "sasl-gsapi", "tls-client-auth", "sasl-md5"....
+        String[] extensions; // "starttls"
+    }
+    List<Endpoint> endpoints;
+    Map<String, String> properties; // additional properties
+}
+
+```
+
+In RegistrationClient interface we expose a new method:
+
+```
+CompletableFuture<Versioned<BookieServiceInfo>> getBookieServiceInfo(String bookieId)
+```
+
+The client can derive bookieId from a BookieSocketAddress. He can access the list of available bookies using **RegistrationClient#getAllBookies()** and then use this new method to get the details of the services exposed by the Bookie.
+
+On the Bookie side we change the RegistrationManager interface in order to let the Bookie
+advertise the services:
+
+in RegistrationManager class the **registerBookie** method 
+```
+void registerBookie(String bookieId, boolean readOnly)
+```
+
+becomes
+
+```
+void registerBookie(String bookieId, boolean readOnly, BookieServiceInfo bookieServieInfo)
+```
+
+It will be up to the implementation of RegistrationManager and RegistrationClient to serialize
+the BookieServiceInfo structure.
+
+For the ZooKeeper based implementation we are going to store such information in JSON format.
+
+```
+{
+    "endpoints": [
+        {
+         "name": "bookie",
+         "hostname": "localhost",
+         "port": 3181,
+         "protocol": "bookie-rpc",
+         "auth": ["sasl-gsapi"],
+         "extensions": ["starttls"]
+        },
+        {
+         "name": "bookie-http-server",
+         "hostname": "localhost",
+         "port": 8080,
+         "protocol": "http",
+         "auth": [],
+         "extensions": []
+        }
+    ],
+    properties {
+        "property": "value",
+        "property2": "value2"
+    }
+}
+```
+Such information will be stored inside the '/REGISTRATIONPATH/available' znode (or /REGISTRATIONPATH/available/readonline' in case of readonly bookie), these paths are the same used in 4.10, but in 4.10 we are writing empty z-nodes.
+
+The rationale around this choice is that the client is already using these znodes in order to discover available bookies services.
+
+It is possible that such endpoint information change during the lifetime of a Bookie, like after rebooting a machine and the machine gets a new network address.
+
+It is out of the scope of this proposal to change semantics of ledger metadata, in which we  are currently storing directly the network address of the bookies, but with this proposal we are preparing the way to have an indirection and separate the concept of Bookie ID from the Network address.
```

Comment:
We can split the work in two tasks.

Currently I am working on this very first task because I need bookies to advertise their capabilities on the discovery service as in my company we are building a BookKeeper Management system.

To me the very important point is to be aware of future directions and prepare our way for the separation of **BookieID** from **Bookie Network Address.**

I will be happy to follow up and work on the second part.
I am already trying to see how client side components like the PlacementPolicy will be in the second iteration.